pkg/agent/plugin/nodeattestor/k8s/k8s.go

```diff
@@ -0,0 +1,245 @@
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/certificate/csr"
+)
+
+const (
+	pluginName = "k8s"
+)
+
+type configData struct {
+	spiffeID        string
+	privateKey      crypto.PrivateKey
+	attestationData *common.AttestationData
+}
+
+type K8sConfig struct {
+	TrustDomain     string `hcl:"trust_domain"`
+	PrivateKeyPath  string `hcl:"k8s_private_key_path"`
+	CertificatePath string `hcl:"k8s_certificate_path"`
+	K8sCACertPath   string `hcl:"k8s_ca_certificate_path"`
+	KubeconfigPath  string `hcl:"kubeconfig_path"`
+}
+
+type K8sPlugin struct {
+	m          sync.Mutex
+	c          *K8sConfig
+	kubeClient kubernetes.Interface
+}
+
+var _ nodeattestor.Plugin = (*K8sPlugin)(nil)
+
+func (p *K8sPlugin) getConfig() *K8sConfig {
+	p.m.Lock()
+	defer p.m.Unlock()
+	return p.c
+}
+
+func (p *K8sPlugin) setConfig(c *K8sConfig) {
+	p.m.Lock()
+	defer p.m.Unlock()
+	p.c = c
+}
+
+func getAgentName() string {
+	name, err := os.Hostname()
+	if err != nil {
+		name = "unknown"
+	}
+	return name
+}
+
+func getKubeClient(kubeConfigFilePath, clientCertFilePath, clientKeyFilePath, caCertFilePath string) (kubernetes.Interface, error) {
+	if kubeConfigFilePath == "" {
+		// Try KUBECONFIG env variable
+		kubeConfigFilePath = os.Getenv("KUBECONFIG")
+		if kubeConfigFilePath == "" {
+			// Still no luck, try default (home)
+			home := os.Getenv("HOME")
+			if home != "" {
+				kubeConfigFilePath = path.Join(home, ".kube", "config")
+			}
+		}
+	}
+
+	if kubeConfigFilePath == "" {
+		return nil, fmt.Errorf("Unable to locate kubeconfig")
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", kubeConfigFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("Error accessing kubeconfig %s: %v", kubeConfigFilePath, err)
+	}
+
+	config.TLSClientConfig.CertFile = clientCertFilePath
+	config.TLSClientConfig.KeyFile = clientKeyFilePath
+	config.TLSClientConfig.CAFile = caCertFilePath
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating clientset: %v", err)
+	}
+	return clientset, nil
+}
+
+func fetchK8sCert(kubeClient kubernetes.Interface) (*tls.Certificate, error) {
+	key, err := cert.MakeEllipticPrivateKeyPEM()
+	if err != nil {
+		return nil, fmt.Errorf("Error creating private key: %v", err)
+	}
+
+	certsIntf := kubeClient.CertificatesV1beta1().CertificateSigningRequests()
+	cert, err := csr.RequestNodeCertificate(certsIntf, key, types.NodeName(getAgentName()))
+	if err != nil {
+		return nil, fmt.Errorf("Error getting certificate: %v", err)
+	}
+
+	tlsCert, err := tls.X509KeyPair(cert, key)
+	if err != nil {
+		return nil, fmt.Errorf("Error forming x509 key pair: %v", err)
+	}
+	return &tlsCert, nil
+}
+
+func (p *K8sPlugin) loadConfigData() (*configData, error) {
+	config := p.getConfig()
+	if config == nil {
+		return nil, errors.New("k8s: not configured")
```

Comment:
nit: might be a good idea to be more specific with these errors, like `k8s node attestor: not configured` or something

Reply:
will do